File: boatpos-server/boatpos-server-repository/boatpos-server-repository-api/src/main/java/org/boatpos/repository/api/model/Rental.java
package org.boatpos.repository.api.model;

import org.boatpos.common.repository.api.model.DomainModel;
import org.boatpos.common.repository.api.values.Deleted;
import org.boatpos.model.PaymentMethod;
import org.boatpos.model.RentalEntity;
import org.boatpos.repository.api.values.*;
import org.boatpos.service.api.bean.RentalBean;

import java.util.Optional;
import java.util.Set;

/**
 * The domain model for a rental.
 */
public interface Rental extends DomainModel<Rental, RentalEntity, RentalBean> {

    /**
     * Undo the delete the current {@link DomainModel}.
     */
    void undoDelete();

    DayId getDayId();

    Rental setDayId(DayId dayId);

    Day getDay();

    Rental setDay(Day day);

    Boat getBoat();

    Rental setBoat(Boat boat);

    DepartureTime getDepartureTime();

    Rental setDepartureTime(DepartureTime departureTime);

    ArrivalTime getArrivalTime();

    Rental setArrivalTime(ArrivalTime arrivalTime);

    PriceCalculatedBefore getPriceCalculatedBefore();

    Rental setPriceCalculatedBefore(PriceCalculatedBefore priceCalculatedBefore);

    PriceCalculatedAfter getPriceCalculatedAfter();

    Rental setPriceCalculatedAfter(PriceCalculatedAfter priceCalculatedAfter);

    PricePaidAfter getPricePaidAfter();

    Rental setPricePaidAfter(PricePaidAfter pricePaidAfter);

    PricePaidBefore getPricePaidBefore();

    Rental setPricePaidBefore(PricePaidBefore pricePaidBefore);

    Finished isFinished();

    Rental setFinished(Finished finished);

    Deleted isDeleted();

    Rental setDeleted(Deleted deleted);

    PaymentMethod getPaymentMethodBefore();

    Rental setPaymentMethodBefore(PaymentMethod paymentMethodBefore);

    PaymentMethod getPaymentMethodAfter();

    Rental setPaymentMethodAfter(PaymentMethod paymentMethodAfter);

    Optional<Promotion> getPromotion();

    Rental setPromotion(Promotion promotion);

    Set<Commitment> getCommitments();

    Rental addCommitments(Set<Commitment> commitments);

    Rental addCommitment(Commitment commitment);

    Rental clearCommitments();

    PricePaidComplete getPricePaidComplete();
}

File: boatpos-client/src/rental.ts
import {Boat} from "./boat";
import {Commitment} from "./commitment";
import {PromotionBefore, PromotionAfter} from "./promotion";

export class Rental {

    dayId:number;
    day:Date;
    boat:Boat;
    departure:Date;
    arrival:Date;
    pricePaidAfter:number;
    pricePaidBefore:number;
    priceCalculatedAfter:number;
    priceCalculatedBefore:number;
    finished:boolean;
    deleted:boolean;
    coupon:boolean;
    promotionBefore:PromotionBefore;
    promotionAfter:PromotionAfter;
    commitments:Array<Commitment> = [];
    timeOfTravel:number;
    timeOfTravelCalculated:number;

    constructor(dayId:number,
                day:Date,
                boat:Boat,
                departure:Date,
                arrival:Date,
                pricePaidAfter:number,
                pricePaidBefore:number,
                priceCalculatedAfter:number,
                priceCalculatedBefore:number,
                finished:boolean,
                deleted:boolean,
                coupon:boolean,
                promotionBefore:PromotionBefore,
                promotionAfter:PromotionAfter,
                commitments:Array<Commitment>,
                timeOfTravel:number,
                timeOfTravelCalculated:number) {
        this.dayId = dayId;
        this.day = day;
        this.boat = boat;
        this.departure = departure;
        this.arrival = arrival;
        this.pricePaidAfter = pricePaidAfter;
        this.pricePaidBefore = pricePaidBefore;
        this.priceCalculatedAfter = priceCalculatedAfter;
        this.priceCalculatedBefore = priceCalculatedBefore;
        this.finished = finished;
        this.deleted = deleted;
        this.coupon = coupon;
        this.promotionBefore = promotionBefore;
        this.promotionAfter = promotionAfter;
        if (commitments != null) {
            commitments.forEach(commitment => this.commitments.push(commitment));
        }
        this.timeOfTravel = timeOfTravel;
        this.timeOfTravelCalculated = timeOfTravelCalculated;
    }

    static fromDeparte(dayId:number,
                       day:Date,
                       boat:Boat,
                       departure:Date,
                       commitments:Array<Commitment>,
                       promotionBefore:PromotionBefore,
                       coupon:boolean,
                       priceCalculatedBefore:number):Rental {
        return new Rental(dayId,
            day,
            boat,
            departure,
            null,
            null,
            null,
            null,
            priceCalculatedBefore,
            null,
            null,
            coupon,
            promotionBefore,
            null,
            commitments,
            null,
            null
        );
    }
}